File: Time-series analysis/trophic_coherence.R
library(matlib) 
library(pracma)
library(igraph)

trophic_coherence <- function(W){
  N = dim(W)[1]
  col_sum = colSums(W)
  row_sum = colSums(t(W))
  u = col_sum+row_sum;
  v = col_sum-row_sum;
  Lambda = diag(u)-W-t(W);
  #Place constant equation for first node
  eq1 = numeric(N)
  eq1[1] = 1
  Lambda[1,] = eq1
  v[1] = 0
  h = solve(Lambda,v)
  num =0
  denom =0
  for (i in 1:N){
    for (j in 1:N){
      num = num + W[i,j]*(h[j]-h[i]-1)^2
      denom = denom + W[i,j]
    }
  }
  return(list(oneminusF0 = 1-num/denom, h = h))
}

trophic_coherence_dg <- function(W){
  N = dim(W)[1]
  col_sum = colSums(W)
  row_sum = colSums(t(W))
  u = col_sum+row_sum;
  v = col_sum-row_sum;
  Lambda = diag(u)-W-t(W);
  
  # Find connected components and modify the equation for one node from each cc
  #cc_info = sort_connected_components(W)
  #no_cc = cc_info$no_cc
  #cc = cc_info$cc
  c = components(graph_from_adjacency_matrix(1*(W>0),mode="directed"), mode="weak")
  no_cc = c$no
  cc = c$membership
  
  # Modify the linear system to guarantee a unique solution 
  for (n in (1:no_cc)){
    found=FALSE
    check = 1
    while (!found){
      if (cc[check] == n){
        eq1 = numeric(N)
        eq1[check] = 1
        Lambda[check,] = eq1
        v[check] = 0
        found = TRUE
      }
      check = check+1
    }
  }
  h = solve(Lambda,v)
  for (n in (1:no_cc)){
    mask = cc==n
    h[mask] = h[mask]-min(h[mask])
  }
  num =0
  denom =0
  for (i in 1:N){
    for (j in 1:N){
      num = num + W[i,j]*(h[j]-h[i]-1)^2
      denom = denom + W[i,j]
    }
  }
  return(list(oneminusF0 = 1-num/denom, h = h))
}

sort_connected_components<-function(W){
  N = dim(W)[1]
  L = diag(rowSums(W))-W
  ns = nullspace(L)
  no_cc = dim(ns)[2]
  cc = kmeans(ns[,1],no_cc)$cluster
  return(list(cc=cc,no_cc = no_cc))
}

File: Network Data Analysis/trophic_coherence.m
function [F,h] = trophic_coherence(W)
[N,~] = size(W);
col_sum = sum(W); row_sum = sum(W');
u = col_sum+row_sum;
v = col_sum-row_sum;
Lambda = diag(u)-W-W';

% modify system to have a unique solution

% % calculate traditional laplacian 
% Laplace = diag(sum((W>0),2)) - (W>0);
% % nullity is number of connected components; basis vectors are constant on
% % CC
% null_basis =null(Laplace);
% [~,no_cc] = size(null_basis);
% % use kmeans (without actually clustering - using exact number of states) to sort the vector into
% % connected components
% %connected_components = kmeans(null_basis(:,1),no_cc);
connected_components = conncomp(graph(0.5*(W+W')));
no_cc = length(unique(connected_components));

for n=1:no_cc
        % find first node in a connected component and modify its equation
        firsteq_cc = find(connected_components==n,1);
        Lambda(firsteq_cc,:) = zeros(N,1);
        Lambda(firsteq_cc,firsteq_cc) = 1;
        v(firsteq_cc) = 1;
end

% solve system for unique solution
h = Lambda\v';

% normalise lowest level in each cc to be 0
for n=1:no_cc
    h(connected_components==n) = h(connected_components==n)-min(h(connected_components==n));
end

num = 0;
denom =sum(W-diag(diag(W)),'all');
for i=1:N
    for j=1:N
        if i~=j
            num = num + W(i,j)*((h(j)-h(i)-1)^2);
        end
    end
end
F = num/denom;
end